### PR TITLE
Replace stableStringify() with canonicalHash()-based stableHash()

### PR DESCRIPTION
**Pull request by danfuzz**
## Summary

Replaces `stableStringify()` with `stableHash()`, a wrapper around `canonicalHash()` from the memory package, for all cache key computation and value deduplication in `traverse.ts` and `cache.ts`.

- **What**: `stableStringify()` (deterministic JSON serialization for cache keys) is replaced by `stableHash()`, which delegates to `canonicalHash().toString()` and uses the same WeakMap identity cache pattern for O(1) repeated hashing. Includes a `_legacyStringify()` fallback for `Date` and `RegExp` edge cases that `canonicalHash` doesn't handle natively (not encountered in practice, but included for safety). All ~15 call sites in `traverse.ts` and 1 in `cache.ts` are updated.
- **Why**: `stableStringify()` cannot handle `bigint` values, blocking bigint flow through the schema/traversal layer. `canonicalHash()` handles all storable types including bigint. This also provides collision-resistant hashing (SHA-256) vs. the previous deterministic string concatenation approach.
- Adds `"./canonical-hash"` to the memory package's `deno.json` exports.

## Files changed (3 files, +76/-61 lines)

- `packages/runner/src/traverse.ts` -- replaced `stableStringify()` with `stableHash()` wrapper; updated all call sites
- `packages/runner/src/storage/cache.ts` -- updated import and single call site
- `packages/memory/deno.json` -- added `"./canonical-hash"` export

## Test plan

- [x] `deno fmt --check` passes on both `packages/runner` and `packages/memory`
- [x] `deno task test` passes for `packages/memory` (64 tests, 0 failures)
- [x] `deno task test` passes for `packages/runner` (201 passed, 1 pre-existing flaky timing test failure in `scheduler-timing.test.ts` unrelated to this change)
- [ ] CI passes

Co-Authored-By: coder-cedar (Claude Opus 4.6) <noreply@anthropic.com>

---
-- upstream-liaison-bolt (Claude Opus 4.6)

### CODE REVIEW

File: packages/runner/src/traverse.ts [hunk 1]
```diff
@@ -83,16 +85,61 @@ const enum TypeValidity {
...
 /**
- * Produces a canonical string representation for use as a hash key.
- * Object keys are sorted for deterministic output so structurally-equal
- * objects always hash identically. Results are cached per object identity
- * via WeakMap, so repeated hashing of the same schema object is O(1).
+ * Produces a canonical hash string for use as a cache/dedup key.
+ * Delegates to `canonicalHash()` (SHA-256 based) with a WeakMap identity
+ * cache so repeated hashing of the same object reference is O(1).
+ *
+ * Falls back to a deterministic JSON-like stringification for values that
```
Comment (danfuzz) [inline packages/runner/src/traverse.ts]: Please add a `TODO(danfuzz)` to remove the fallbacks once the data model code grows to cover these cases directly.

Comment (danfuzz) [reply packages/runner/src/traverse.ts]: Added `// TODO(danfuzz): Remove fallbacks once the data model covers NaN, Infinity, Date, and RegExp directly.` above the `_stableHashCache` declaration.

-- upstream-liaison-bolt (Claude Opus 4.6)

File: packages/runner/src/traverse.ts [hunk 2]
```diff
@@ -83,16 +85,61 @@ const enum TypeValidity {
...
+ * `canonicalHash` rejects (NaN, Infinity, Date, RegExp). These don't
+ * appear in normal JSON Schema data but may surface in edge-case tests.
  */
-const _hashCache = new WeakMap<object, string>();
+const _stableHashCache = new WeakMap<object, string>();
+
+export function stableHash(value: unknown): string {
+  if (value === null) return "h:null";
```
[the remark this answers is not in the document]
Comment (danfuzz) [reply packages/runner/src/traverse.ts]: Good catch -- no, there's no problem with how `canonicalHash()` handles them. Both `null` and `undefined` are covered by `canonicalHash`'s primitive constants cache (from PR #2986). The early returns were carried over from the original `stableStringify` implementation but aren't needed. Removed them -- `null` and `undefined` now fall through to `canonicalHash(value).toString()`.

-- upstream-liaison-bolt (Claude Opus 4.6)
